### ISSUE TITLE
Reducing share permissions doesn't propagate over reshares

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#13201]
### Steps to reproduce
1. Create three users: "user1", "user2", "user3"
2. "user1" shares a folder "test" with "user2" with all permissions
3. "user2" reshares "test" with "user3" with all permissions
4. "user1" removes the "update" permission from "test"
5. "user3" uploads a file into "test"

### Expected result
"user3" cannot upload to "test", as the permissions have been reduced.

### Actual result
"user3" can STILL upload to "test"

### Versions
ownCloud 8 a5099b01f93ec81602c8facba04233c7bc65ed47

CC @schiesbn @icewind1991 @jnfrmarks

**Comment by PVince81** [comment]
This is not a regression, OC 6 and OC 7 also have the issue.
According to @schiesbn this issue likely has existed since the dawn of the sharing feature.

If we decide on introducing single list shares (https://github.com/owncloud/core/issues/9058) such issues would be obsolete.

**Comment by PVince81** [comment]
I have the feeling that https://github.com/owncloud/core/issues/9058 won't be scheduled any soon...
I'll see if there's a possible quickfix for this.

**Comment by PVince81** [comment]
Wow... this seems to work now on master (a65f666834742f8df25493a11e10063e1e5ff98a)
Permissions do propagate.

I also tried to have u2 share a "subfolder" from "test" with u3 instead of the main folder: permissions also propagate properly.

Also works if at the end of the chain a public link was created with allow upload: the permission is also removed when the parent loses its permission.

Closing as fixed.
